client/internal/clientcommon.go

```diff
@@ -72,30 +70,68 @@ func NewClientCommon(logger types.Logger, sender Sender) ClientCommon {
 	return ClientCommon{Logger: logger, sender: sender, stoppedSignal: make(chan struct{}, 1)}
 }
 
+func (c *ClientCommon) hasCapability(capability protobufs.AgentCapabilities) bool {
+	return c.ClientSyncedState.Capabilities()&capability != 0
+}
+
+func (c *ClientCommon) validateCapabilities(capabilities protobufs.AgentCapabilities) error {
+	if capabilities&protobufs.AgentCapabilities_AgentCapabilities_ReportsHealth != 0 && c.ClientSyncedState.Health() == nil {
+		return ErrHealthMissing
+	}
+	if capabilities&protobufs.AgentCapabilities_AgentCapabilities_ReportsAvailableComponents != 0 && c.ClientSyncedState.AvailableComponents() == nil {
+		return ErrAvailableComponentsMissing
+	}
+	if c.PackagesStateProvider != nil {
+		if (capabilities&protobufs.AgentCapabilities_AgentCapabilities_AcceptsPackages == 0) ||
+			(capabilities&protobufs.AgentCapabilities_AgentCapabilities_ReportsPackageStatuses == 0) {
+			return ErrAcceptsPackagesNotSet
+		}
+	} else {
+		if capabilities&protobufs.AgentCapabilities_AgentCapabilities_AcceptsPackages != 0 ||
+			capabilities&protobufs.AgentCapabilities_AgentCapabilities_ReportsPackageStatuses != 0 {
+			return ErrPackagesStateProviderNotSet
+		}
+	}
+	return nil
+}
+
 // PrepareStart prepares the client state for the next Start() call.
 // It returns an error if the client is already started, or if the settings are invalid.
 func (c *ClientCommon) PrepareStart(
-	_ context.Context, settings types.StartSettings,
+	ctx context.Context, settings types.StartSettings,
 ) error {
 	if c.isStarted {
 		return errAlreadyStarted
 	}
-
-	c.Capabilities = settings.Capabilities
-
-	// According to OpAMP spec this capability MUST be set, since all Agents MUST report status.
-	c.Capabilities |= protobufs.AgentCapabilities_AgentCapabilities_ReportsStatus
+	// Deprecated: Use client.SetCapabilities() instead.
+	if settings.Capabilities != 0 {
+		c.Logger.Errorf(ctx, "settings.Capabilities is deprecated, use client.SetCapabilities() instead")
+		capabilities := settings.Capabilities
+		// According to OpAMP spec this capability MUST be set, since all Agents MUST report status.
+		capabilities |= protobufs.AgentCapabilities_AgentCapabilities_ReportsStatus
+		if err := c.ClientSyncedState.SetCapabilities(&capabilities); err != nil {
+			return err
```

Comment:
@tigrannajaryan logically, this state can never happen but codecov isn't smart enough to realize that... the only way setcapabilities could error is if the capabilities passed in here is nil, but capabilities is a reference to an instantiated variable in both this case and the next case which means this could never happen. How should I proceed here?

Reply:
That's fine, ignore codecov.